test/unittest/tests/block/test_block_session.cpp

```diff
@@ -0,0 +1,623 @@
+
+        REQUIRE(__wti_block_ext_alloc(session->getWtSessionImpl(), &ext) == 0);
+        // Construct extent cache with one item.
+        bms->ext_cache = ext;
+        bms->ext_cache_cnt = 0;
+
+        REQUIRE(__wti_block_ext_alloc(session->getWtSessionImpl(), &cached_ext) == 0);
+        REQUIRE(cached_ext == ext);
+        validate_and_cleanup_ext_block(ext);
+    }
+
+    SECTION("Allocate with one extent in cache")
+    {
+        WT_EXT *ext;
+        WT_EXT *cached_ext;
+
+        REQUIRE(__wti_block_ext_alloc(session->getWtSessionImpl(), &ext) == 0);
+        // Construct extent cache with one item.
+        bms->ext_cache = ext;
+        bms->ext_cache_cnt = 1;
+
+        REQUIRE(__wti_block_ext_alloc(session->getWtSessionImpl(), &cached_ext) == 0);
+        REQUIRE(cached_ext == ext);
+        validate_and_cleanup_ext_block(ext);
+    }
+
+    SECTION("Allocate with two extents in cache ")
+    {
+        WT_EXT *ext;
+        WT_EXT *ext2;
+        WT_EXT *cached_ext;
+
+        REQUIRE(__wti_block_ext_alloc(session->getWtSessionImpl(), &ext) == 0);
+        REQUIRE(__wti_block_ext_alloc(session->getWtSessionImpl(), &ext2) == 0);
+
+        // Construct extent cache with two items.
+        ext->next[0] = ext2;
+        bms->ext_cache = ext;
+        bms->ext_cache_cnt = 2;
+
+        REQUIRE(__wti_block_ext_alloc(session->getWtSessionImpl(), &cached_ext) == 0);
+        REQUIRE(ext == cached_ext);
+        REQUIRE(ext2 != cached_ext);
+        validate_and_cleanup_ext_list(bms, 1);
+    }
+}
+
+TEST_CASE("Block session: __wti_block_ext_free", "[block_session]")
+{
+    std::shared_ptr<MockSession> session = MockSession::buildTestMockSession();
+    WT_BLOCK_MGR_SESSION *bms = session->setupBlockManagerSession();
+
+    SECTION("Free with null block manager session -- needs discussion")
+    {
+        std::shared_ptr<MockSession> session_no_bm = MockSession::buildTestMockSession();
+        WT_EXT *ext;
+
+        REQUIRE(__ut_block_ext_alloc(session_no_bm->getWtSessionImpl(), &ext) == 0);
+        REQUIRE(ext != nullptr);
+
+        __wti_block_ext_free(session_no_bm->getWtSessionImpl(), ext);
+        // REQUIRE(ext == nullptr);
+    }
+
+    SECTION("Calling free with cache")
+    {
+        WT_EXT *ext;
+        WT_EXT *ext2;
+
+        REQUIRE(__ut_block_ext_alloc(session->getWtSessionImpl(), &ext) == 0);
+
+        __wti_block_ext_free(session->getWtSessionImpl(), ext);
+
+        REQUIRE(ext != nullptr);
+        REQUIRE(bms->ext_cache == ext);
+        validate_ext_list(bms, 1);
+
+        REQUIRE(__ut_block_ext_alloc(session->getWtSessionImpl(), &ext2) == 0);
+        __wti_block_ext_free(session->getWtSessionImpl(), ext2);
+
+        REQUIRE(ext != nullptr);
+        REQUIRE(bms->ext_cache == ext2);
+        REQUIRE(bms->ext_cache->next[0] == ext);
+        validate_and_cleanup_ext_list(bms, 2);
+    }
+}
+
+TEST_CASE("Block session: __block_ext_discard", "[block_session]")
+{
+    WT_EXT *ext, *ext2, *ext3;
```

Comment:
I suggest moving the allocation of `ext`, `ext2` and `ext3` to just above the line where `ext` is first used.